tornado/testing.py

```diff
@@ -499,13 +506,16 @@ def wrap(f):
         @functools.wraps(f)
         def pre_coroutine(self, *args, **kwargs):
             result = f(self, *args, **kwargs)
-            if isinstance(result, GeneratorType):
+            if isinstance(result, GeneratorType) or iscoroutine(result):
                 self._test_generator = result
             else:
                 self._test_generator = None
             return result
 
-        coro = gen.coroutine(pre_coroutine)
+        if iscoroutinefunction(f):
```

Comment:
Apparently coroutines are *not* a subtype of generators:
```
from collections.abc import Generator, Coroutine

def gen_fn():
    yield

async def coro_fn():
    pass

print(isinstance(gen_fn(), Generator))   # True
print(isinstance(coro_fn(), Generator))  # False
print(isinstance(coro_fn(), Coroutine))  # True
```
But in any case you're right, `gen.coroutine` can wrap either. I'll simplify this.